src/sentry/monitors/tasks/detect_broken_monitor_envs.py

```diff
@@ -41,40 +60,104 @@ def detect_broken_monitor_envs():
-                id=open_incident.monitor.organization_id
-            )
+            organization = Organization.objects.get_from_cache(id=org_id)
             if not features.has(
                 "organizations:crons-broken-monitor-detection", organization=organization
             ):
                 continue
         except Organization.DoesNotExist:
             continue
 
-        # verify that the most recent check-ins have been failing
-        recent_checkins = (
-            MonitorCheckIn.objects.filter(monitor_environment=open_incident.monitor_environment)
-            .order_by("-date_added")
-            .values("status")[:NUM_CONSECUTIVE_BROKEN_CHECKINS]
+        # Map user email to a dictionary of monitors and their earliest incident start date amongst its broken environments
+        user_broken_envs = defaultdict(
+            lambda: defaultdict(
+                lambda: {"environment_names": [], "earliest_start": django_timezone.now()}
+            )
         )
-        if len(recent_checkins) != NUM_CONSECUTIVE_BROKEN_CHECKINS or not all(
-            checkin["status"] in [CheckInStatus.ERROR, CheckInStatus.TIMEOUT, CheckInStatus.MISSED]
-            for checkin in recent_checkins
+        orgs_open_incidents = (
+            open_incidents_qs.all()
+            .select_related("monitor_environment")
+            .filter(monitor__organization_id=org_id)
+        )
+        # Query for all the broken incidents within the current org we are processing
+        for open_incident in RangeQuerySetWrapper(
+            orgs_open_incidents,
+            order_by="starting_timestamp",
+            step=1000,
         ):
-            continue
+            # Verify that the most recent check-ins have been failing
+            recent_checkins = (
+                MonitorCheckIn.objects.filter(monitor_environment=open_incident.monitor_environment)
+                .order_by("-date_added")
+                .values("status")[:NUM_CONSECUTIVE_BROKEN_CHECKINS]
+            )
+            if len(recent_checkins) != NUM_CONSECUTIVE_BROKEN_CHECKINS or not all(
+                checkin["status"]
+                in [CheckInStatus.ERROR, CheckInStatus.TIMEOUT, CheckInStatus.MISSED]
+                for checkin in recent_checkins
+            ):
+                continue
 
-        detection, _ = MonitorEnvBrokenDetection.objects.get_or_create(
-            monitor_incident=open_incident, defaults={"detection_timestamp": current_time}
-        )
-        if not detection.user_notified_timestamp:
-            # TODO(davidenwang): This is where we would implement email sending logic
-            pass
+            detection, _ = MonitorEnvBrokenDetection.objects.get_or_create(
+                monitor_incident=open_incident, defaults={"detection_timestamp": current_time}
+            )
+            if not detection.user_notified_timestamp:
+                environment_name = open_incident.monitor_environment.get_environment().name
```

Comment:
n+1 here, but not sure how much we really care about this.

Does `get_environment` do some cache lookup?

Reply:
yea it does `get_by_cache`. I guess would the alternative be to fetch all environments for the org according to the environment ids inside of all the incidents in the org?